lib/device_detector.rb

```diff
@@ -44,7 +41,55 @@ def device_name
   end
 
   def device_type
-    device.type
+    t = device.type
+
+    if t.nil? && android_tablet_fragment?
+      t = "tablet"
+    end
+
+    if t.nil? && android_mobile_fragment?
+      t = "smartphone"
+    end
+
+    # Android up to 3.0 was designed for smartphones only. But as 3.0,
+    # which was tablet only, was published too late, there were a
+    # bunch of tablets running with 2.x With 4.0 the two trees were
+    # merged and it is for smartphones and tablets
+    #
+    # So were are expecting that all devices running Android < 2 are
+    # smartphones Devices running Android 3.X are tablets. Device type
+    # of Android 2.X and 4.X+ are unknown
+    if t.nil? && os.short_name == 'AND' && os.full_version && !os.full_version.empty?
+      if os.full_version < "2"
+        t = "smartphone"
+      elsif os.full_version >= "3" && os.full_version < "4"
+        t = "tablet"
+      end
+    end
+
+    # All detected feature phones running android are more likely a smartphone
+    if t == "feature phone" && os.family == 'Android'
+      t = "smartphone"
+    end
+
+    # According to http://msdn.microsoft.com/en-us/library/ie/hh920767(v=vs.85).aspx
+    # Internet Explorer 10 introduces the "Touch" UA string token. If this token is present at the end of the
+    # UA string, the computer has touch capability, and is running Windows 8 (or later).
+    # This UA string will be transmitted on a touch-enabled system running Windows 8 (RT)
+    #
+    # As most touch enabled devices are tablets and only a smaller part are desktops/notebooks we assume that
+    # all Windows 8 touch devices are tablets.
+    if t.nil? && touch_enabled? &&
+       (os.short_name == 'WRT' || (os.short_name == 'WIN' && os.full_version && os.full_version >= '8'))
+      t = "tablet"
+    end
+
+    # set device type to desktop for all devices running a desktop os that were not detected as an other device type
+    if t.nil? && os.desktop?
+      t = "desktop"
+    end
```

Comment:
I think that the type detecting logic does belong in the `Device#type` method. The file `lib/device_detector.rb` is more of a proxy / public interface of the gem.